conduit/test/Spec.hs

```diff
@@ -652,6 +655,44 @@ spec = do
         res1 <- runConduit $ yieldMany strs .| linesUnboundedC .| sinkList
         res2 <- runConduit $ yieldMany strs .| peekForeverE (lineC $ foldC >>= yield) .| sinkList
         res2 `shouldBe` res1
+    describe "concurrentMap" $ do
+        let boom = stringException "boom"
+        let isBoom e = maybe False (== boom) $ fromException e
+        let isLinkedBoom e = case fromException e of
+                Nothing -> False
+                Just (ExceptionInLinkedThread _ inner) -> isBoom inner
+        let block = newEmptyMVar >>= takeMVar :: IO ()
+        it "propagates exceptions" $ do
+            let go = runConduitRes $ yield () .| concurrentMap 10 (\_ -> throwM boom) .| sinkNull
+            go `shouldThrow` isLinkedBoom
```

Comment:
This test is flaky and I haven't figured out why yet.

<img width="565" alt="image" src="https://user-images.githubusercontent.com/1387653/154910728-492a8bbe-da17-411a-b66e-9cea4c5248a1.png">

I verified `isLinkedBoom` returns `True` by wrapping it:

    go `catch` (\e -> print ("isLinkedBoom?", isLinkedBoom e) >> throwM e)

This one's got me stumped.

Reply:
I think it was caused by some race condition that would trigger 2 exceptions. The first exception would get caught, checked, and handled properly, but the second would strike outside the scope of the exception handlers (before or after).

I removed the `link worker` call (it's safe because all workers are `poll`ed, `wait`ed, or `cancel`ed), which means exceptions only bubble up when execution on the main thread is inside `concurrentMap` (e.g. if the main thread is in the source when a worker hits an exception then the main thread wouldn't know about it until the main thread yields a value from the source and `concurrentMap` reads it). I also tweaked the test to take that into account.